pixeltable/exprs/expr.py

```diff
@@ -699,78 +707,108 @@ def __rfloordiv__(self, other: object) -> 'exprs.ArithmeticExpr':
+                )
+            other = other_lit
+        return StringOp(op, other, self)
 
     def _make_arithmetic_expr(self, op: ArithmeticOperator, other: object) -> 'exprs.ArithmeticExpr':
-        """
-        other: Union[Expr, LiteralPythonTypes]
-        """
-        # TODO: check for compatibility
         from .arithmetic_expr import ArithmeticExpr
         from .literal import Literal
 
-        if isinstance(other, Expr):
-            return ArithmeticExpr(op, self, other)
-        if isinstance(other, typing.get_args(LiteralPythonTypes)):
-            return ArithmeticExpr(op, self, Literal(other))
-        raise TypeError(f'Other must be Expr or literal: {type(other)}')
+        if not isinstance(other, Expr):
+            other_lit = Literal.from_object(other)
+            if other_lit is None or not other_lit.col_type.is_numeric_type():
+                op_symbol = str(op)
+                raise TypeError(
+                    f'ERROR during arithmetic operation ({op_symbol}): Operand must be an '
+                    'Expression or a numeric literal, '
+                    f'but received type {{type(other).__name__}}.'
+                )
+            other = other_lit
+        return ArithmeticExpr(op, self, other)
 
     def _rmake_arithmetic_expr(self, op: ArithmeticOperator, other: object) -> 'exprs.ArithmeticExpr':
-        """
-        Right-handed version of _make_arithmetic_expr. other must be a literal; if it were an Expr,
-        the operation would have already been evaluated in its left-handed form.
-        """
-        # TODO: check for compatibility
         from .arithmetic_expr import ArithmeticExpr
         from .literal import Literal
 
-        assert not isinstance(other, Expr)  # Else the left-handed form would have evaluated first
-        if isinstance(other, typing.get_args(LiteralPythonTypes)):
-            return ArithmeticExpr(op, Literal(other), self)
-        raise TypeError(f'Other must be Expr or literal: {type(other)}')
+        if not isinstance(other, Expr):
+            other_lit = Literal.from_object(other)
+            if other_lit is None or not other_lit.col_type.is_numeric_type():
+                op_symbol = str(op)
+                raise TypeError(
+                    f'ERROR during arithmetic operation ({op_symbol}): Left operand must be an '
+                    'Expression or a numeric literal, '
+                    f'but received type {{type(other).__name__}}.'
+                )
+            other = other_lit
+        return ArithmeticExpr(op, other, self)
 
     def __and__(self, other: object) -> Expr:
+        # TODO: check for compatibility
         if not isinstance(other, Expr):
-            raise TypeError(f'Other needs to be an expression: {type(other)}')
+            raise TypeError(
+                'ERROR in logical AND (&): Right operand must be a Pixeltable Expression, '
+                'but received type {{type(other).__name__}}.'
```

Comment:
More user-facing junk. Unless there's something I'm really not understanding here. But I don't think there is.